File: backend/internal/consumer/escpos/escpos.go
package escpos

import (
	"bytes"
	"fmt"
	"strings"
	"time"
)

type Printer struct {
	buf *bytes.Buffer
}

func NewPrinter() *Printer {
	return &Printer{
		buf: bytes.NewBuffer(nil),
	}
}

func (p *Printer) Bytes() []byte {
	return p.buf.Bytes()
}

func (p *Printer) Reset() {
	p.buf.Reset()
}

func (p *Printer) write(b []byte) {
	p.buf.Write(b)
}

func (p *Printer) Initialize() {
	p.write([]byte{0x1B, 0x40})
}

func (p *Printer) Cut() {
	p.write([]byte{0x1D, 0x56, 0x00})
}

func (p *Printer) CutPartial() {
	p.write([]byte{0x1D, 0x56, 0x01})
}

func (p *Printer) Feed(n int) {
	p.write([]byte{0x1B, 0x64, byte(n)})
}

func (p *Printer) NewLine() {
	p.write([]byte{0x0A})
}

func (p *Printer) SetTextNormal() {
	p.write([]byte{0x1B, 0x21, 0x00})
}

func (p *Printer) SetTextBold(enable bool) {
	if enable {
		p.write([]byte{0x1B, 0x45, 0x01})
	} else {
		p.write([]byte{0x1B, 0x45, 0x00})
	}
}

func (p *Printer) SetTextSize(width, height int) {
	w := byte(width - 1)
	h := byte(height - 1)
	if w > 7 {
		w = 7
	}
	if h > 7 {
		h = 7
	}
	p.write([]byte{0x1D, 0x21, (w << 4) | h})
}

func (p *Printer) SetTextDoubleHeight(enable bool) {
	if enable {
		p.write([]byte{0x1B, 0x21, 0x10})
	} else {
		p.write([]byte{0x1B, 0x21, 0x00})
	}
}

func (p *Printer) SetTextDoubleWidth(enable bool) {
	if enable {
		p.write([]byte{0x1B, 0x21, 0x20})
	} else {
		p.write([]byte{0x1B, 0x21, 0x00})
	}
}

func (p *Printer) SetUnderline(enable bool) {
	if enable {
		p.write([]byte{0x1B, 0x2D, 0x01})
	} else {
		p.write([]byte{0x1B, 0x2D, 0x00})
	}
}

func (p *Printer) SetAlign(align string) {
	var b byte
	switch strings.ToLower(align) {
	case "center":
		b = 0x01
	case "right":
		b = 0x02
	default:
		b = 0x00
	}
	p.write([]byte{0x1B, 0x61, b})
}

func (p *Printer) SetFont(font string) {
	var b byte
	switch strings.ToLower(font) {
	case "b":
		b = 0x01
	default:
		b = 0x00
	}
	p.write([]byte{0x1B, 0x4D, b})
}

func (p *Printer) PrintText(text string) {
	p.write([]byte(text))
}

func (p *Printer) PrintLine(text string) {
	p.PrintText(text)
	p.NewLine()
}

func (p *Printer) PrintLineWithAlign(text string, align string) {
	p.SetAlign(align)
	p.PrintLine(text)
	p.SetAlign("left")
}

func (p *Printer) PrintTitle(text string) {
	p.SetTextBold(true)
	p.SetTextSize(2, 2)
	p.SetAlign("center")
	p.PrintLine(text)
	p.SetTextNormal()
	p.Feed(1)
}

func (p *Printer) PrintSeparator() {
	p.PrintLine("--------------------------------")
}

func (p *Printer) PrintDoubleSeparator() {
	p.PrintLine("================================")
}

func (p *Printer) PrintItem(name string, quantity int, price string, amount string) {
	line := fmt.Sprintf("%-20s %3d %6s %8s", name, quantity, price, amount)
	p.PrintLine(line)
}

func (p *Printer) PrintItemWithWidth(name string, quantity int, price string, amount string, nameWidth int) {
	format := fmt.Sprintf("%%-%ds %%3d %%6s %%8s", nameWidth)
	line := fmt.Sprintf(format, name, quantity, price, amount)
	p.PrintLine(line)
}

func (p *Printer) PrintSummary(label string, value string) {
	line := fmt.Sprintf("%-20s %14s", label, value)
	p.SetTextBold(true)
	p.PrintLine(line)
	p.SetTextBold(false)
}

func (p *Printer) PrintHeader(storeName, orderNo string) {
	p.Initialize()
	p.Feed(2)
	p.PrintTitle(storeName)
	p.PrintLineWithAlign("--------------------------------", "center")
	p.SetAlign("left")
	p.PrintLine(fmt.Sprintf("订单号: %s", orderNo))
	p.PrintLine(fmt.Sprintf("时间: %s", time.Now().Format("2006-01-02 15:04:05")))
	p.PrintSeparator()
	p.Feed(1)
}

func (p *Printer) PrintFooter(footer string) {
	p.Feed(1)
	p.PrintSeparator()
	if footer != "" {
		p.SetAlign("center")
		p.PrintLine(footer)
		p.SetAlign("left")
	}
	p.Feed(3)
	p.Cut()
}

func (p *Printer) PrintBarcode(code string, barcodeType byte) {
	p.write([]byte{0x1D, 0x6B, barcodeType})
	p.write([]byte(code))
	p.write([]byte{0x00})
}

func (p *Printer) PrintQRCode(code string, size byte) {
	if size < 1 {
		size = 8
	}
	if size > 16 {
		size = 16
	}
	p.write([]byte{0x1D, 0x28, 0x6B, 0x03, 0x00, 0x31, 0x43, size})
	p.write([]byte{0x1D, 0x28, 0x6B, 0x03, 0x00, 0x31, 0x45, 0x30})
	codeLen := len(code) + 3
	pl := byte(codeLen % 256)
	ph := byte(codeLen / 256)
	p.write([]byte{0x1D, 0x28, 0x6B, pl, ph, 0x31, 0x50, 0x30})
	p.write([]byte(code))
	p.write([]byte{0x1D, 0x28, 0x6B, 0x03, 0x00, 0x31, 0x51, 0x30})
}

func (p *Printer) OpenCashDrawer() {
	p.write([]byte{0x1B, 0x70, 0x00, 0x3C, 0x78})
}

File: printer-service/src/printers/escpos.js
'use strict';

const iconv = require('iconv-lite');

class EscPosPrinter {
  constructor() {
    this.buffer = Buffer.alloc(0);
    this.encoding = 'UTF-8';
  }

  raw(data) {
    if (typeof data === 'string') {
      data = iconv.encode(data, this.encoding);
    }
    this.buffer = Buffer.concat([this.buffer, data]);
    return this;
  }

  esc(cmd, ...args) {
    this.buffer = Buffer.concat([this.buffer, Buffer.from([0x1b, cmd, ...args])]);
    return this;
  }

  gs(cmd, ...args) {
    this.buffer = Buffer.concat([this.buffer, Buffer.from([0x1d, cmd, ...args])]);
    return this;
  }

  setEncoding(encoding) {
    this.encoding = encoding;
    return this;
  }

  setTextBold(bold = true) {
    return this.esc(0x45, bold ? 0x01 : 0x00);
  }

  setTextSize(width = 1, height = 1) {
    return this.gs(0x21, (width - 1) | ((height - 1) << 4));
  }

  setDoubleWidth(enable = true) {
    return this.setTextSize(enable ? 2 : 1, 1);
  }

  setTextAlign(align = 'left') {
    const alignMap = {
      left: 0x00,
      center: 0x01,
      right: 0x02,
      lt: 0x00,
      ct: 0x01,
      rt: 0x02,
    };
    return this.esc(0x61, alignMap[align] || 0x00);
  }

  align(align) {
    return this.setTextAlign(align);
  }

  style(style) {
    if (style === 'bu') {
      this.setTextBold(true);
    } else if (style === 'normal') {
      this.setTextBold(false);
    }
    return this;
  }

  size(width, height) {
    return this.setTextSize(width, height);
  }

  text(content) {
    return this.raw(content + '\n');
  }

  println(content) {
    return this.text(content);
  }

  PrintLine(content) {
    return this.text(content);
  }

  PrintSeparator(char = '-', len = 32) {
    return this.raw(char.repeat(len) + '\n');
  }

  SetTextBold(bold) {
    return this.setTextBold(bold);
  }

  Feed(lines = 1) {
    this.buffer = Buffer.concat([this.buffer, Buffer.from([0x1b, 0x64, lines])]);
    return this;
  }

  feed(lines = 1) {
    return this.Feed(lines);
  }

  Cut(mode = 'full') {
    const modeVal = mode === 'partial' ? 0x01 : 0x00;
    this.buffer = Buffer.concat([this.buffer, Buffer.from([0x1d, 0x56, modeVal])]);
    return this;
  }

  cut(mode = 'full') {
    return this.Cut(mode);
  }

  close() {
    return this;
  }

  PrintHeader(storeName, orderNo) {
    return this
      .setTextAlign('center')
      .setTextBold(true)
      .setTextSize(2, 2)
      .text(storeName)
      .setTextSize(1, 1)
      .text(`单号: ${orderNo}`)
      .Feed(1)
      .setTextAlign('left')
      .setTextBold(false);
  }

  PrintFooter(footerText) {
    return this
      .Feed(1)
      .setTextAlign('center')
      .text(footerText)
      .Feed(2);
  }

  PrintSummary(label, value) {
    return this
      .setTextBold(true)
      .text(`${label}${value}`)
      .setTextBold(false);
  }

  PrintItemWithWidth(name, qty, price, subtotal, nameWidth = 18) {
    const truncatedName = name.length > nameWidth ? name.slice(0, nameWidth - 3) + '...' : name;
    const line = `${truncatedName.padEnd(nameWidth)} ${String(qty).padStart(3)} ${String(price).padStart(6)} ${String(subtotal).padStart(8)}`;
    return this.text(line);
  }

  PrintQRCode(content, size = 8) {
    if (!content || content.length === 0) {
      return this;
    }

    const data = Buffer.from(content, 'utf8');
    const len = data.length;
    const pL = len & 0xff;
    const pH = (len >> 8) & 0xff;
    const moduleSize = Math.min(Math.max(size, 1), 16);

    this.buffer = Buffer.concat([
      this.buffer,
      Buffer.from([0x1d, 0x28, 0x6b, 0x03, 0x00, 0x31, 0x43, moduleSize]),
      Buffer.from([0x1d, 0x28, 0x6b, 0x03, 0x00, 0x31, 0x45, 0x33]),
      Buffer.from([0x1d, 0x28, 0x6b, pL + 3, pH, 0x31, 0x50, 0x30]),
      data,
      Buffer.from([0x1d, 0x28, 0x6b, 0x03, 0x00, 0x31, 0x51, 0x30]),
    ]);

    return this;
  }

  PrintQRCodeWithText(content, text = '', size = 8) {
    if (text) {
      this.setTextAlign('center');
      this.setTextBold(true);
      this.text(text);
      this.setTextBold(false);
      this.Feed(1);
    }
    this.setTextAlign('center');
    this.PrintQRCode(content, size);
    this.Feed(1);
    this.setTextAlign('left');
    return this;
  }

  async PrintImageFromURL(imageUrl, maxWidth = 384) {
    try {
      const sharp = await import('sharp').catch(() => null);
      if (!sharp) {
        console.warn('[EscPos] sharp 库未安装，跳过图片打印');
        return this;
      }

      const response = await fetch(imageUrl);
      if (!response.ok) {
        console.warn('[EscPos] 图片下载失败:', imageUrl);
        return this;
      }
      const arrayBuffer = await response.arrayBuffer();
      const buffer = Buffer.from(arrayBuffer);

      const { data, info } = await sharp.default(buffer)
        .grayscale()
        .resize(maxWidth, null, { fit: 'inside', withoutEnlargement: true })
        .raw()
        .toBuffer({ resolveWithObject: true });

      return this.PrintBitmap(data, info.width, info.height);
    } catch (err) {
      console.warn('[EscPos] 图片打印失败:', err.message);
      return this;
    }
  }

  PrintBitmap(imageData, width, height) {
    const byteWidth = Math.ceil(width / 8);
    const threshold = 128;

    const bits = [];
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < byteWidth; x++) {
        let byte = 0;
        for (let b = 0; b < 8; b++) {
          const px = x * 8 + b;
          if (px < width) {
            const idx = y * width + px;
            if (imageData[idx] < threshold) {
              byte |= 0x80 >> b;
            }
          }
        }
        bits.push(byte);
      }
    }

    const xL = byteWidth & 0xff;
    const xH = (byteWidth >> 8) & 0xff;
    const yL = height & 0xff;
    const yH = (height >> 8) & 0xff;

    const header = Buffer.from([0x1d, 0x76, 0x30, 0x00, xL, xH, yL, yH]);
    const data = Buffer.from(bits);

    this.buffer = Buffer.concat([this.buffer, header, data]);
    return this;
  }

  Bytes() {
    return this.buffer;
  }

  bytes() {
    return this.buffer;
  }
}

function NewPrinter() {
  return new EscPosPrinter();
}

module.exports = {
  EscPosPrinter,
  NewPrinter,
  Printer: EscPosPrinter,
};
